snuba/split.py

```diff
@@ -9,86 +9,155 @@
+
+            if remaining_offset > 0 and len(overall_result['data']) > 0:
+                to_trim = min(remaining_offset, len(overall_result['data']))
+                overall_result['data'] = overall_result['data'][to_trim:]
+                remaining_offset -= to_trim
+
+            total_results = len(overall_result['data'])
+
+            if total_results < limit:
+                if len(result['data']) == 0:
+                    # If we got nothing from the last query, expand the range by a static factor
+                    split_step = split_step * STEP_GROWTH
+                else:
+                    # If we got some results but not all of them, estimate how big the time
+                    # range should be for the next query based on how many results we got for
+                    # our last query and its time range, and how many we have left to fetch.
+                    remaining = limit - total_results
+                    split_step = split_step * math.ceil(remaining / float(len(result['data'])))
+
+                # Set the start and end of the next query based on the new range.
+                split_end = split_start
+                try:
+                    split_start = max(split_end - timedelta(seconds=split_step), from_date)
+                except OverflowError:
+                    split_start = from_date
+
+        return overall_result, status
+
+    def col_split(*args, **kwargs):
+        """
+        Split query in 2 steps if a large number of columns is being selected.
+            - First query only selects event_id and project_id.
+            - Second query selects all fields for only those events.
+            - If the orderby term is based on timestamp also update the date range
+        """
+        body = args[0]
+        orderby = util.to_list(body.get('orderby'))
+        is_timestamp_ordered = orderby and orderby[0] in ('timestamp', '-timestamp')
+
+        min_cols = ['project_id', 'event_id']
+        if is_timestamp_ordered:
+            min_cols.append('timestamp')
+
+        minimal_query = dict(body, **{'selected_columns': min_cols})
+
+        result, status = query_func(*((minimal_query, ) + args[1:]), **kwargs)
+
+        # If something failed, just return
+        if status != 200:
+            return result, status
+
+        conditions = body.get('conditions', [])
+
+        if result:
+            project_ids = list(set([event['project_id'] for event in result['data']]))
+            conditions.append(('project_id', 'IN', project_ids))
+
+            event_ids = list(set([event['event_id'] for event in result['data']]))
+            conditions.append(('event_id', 'IN', event_ids))
+
+            if is_timestamp_ordered:
+                timestamps = [event['timestamp'] for event in result['data']]
+                from_date = util.parse_datetime(min(timestamps))
+                to_date = util.parse_datetime(max(timestamps)) + timedelta(seconds=1)
```

Comment:
It probably makes sense to at least comment here why this `+ timedelta(seconds=1)` works for events. We should later consider adding something like `{from,to}_date_inclusive` parameters to change the behavior here, since this isn't going to work if anywhere we'd deal with higher precision timestamps.

Reply:
Will add a comment. I think it's worth just changing behavior even since Sentry will also have to implement this workaround in places